### DISCUSSION TITLE
Deployment-specific environment variables

### DISCUSSION THREAD

**Discussion by johnrichardrinehart** [vercel/community#1059] [Help] [answered]
We have a need to deploy different instances of our site, with application environment variables adjusted on a per-deployment basis. Basically, we want an Action to trigger a mirror deployment on merges to `main` that's not pointing to our `production` backend. If we use the `vercel` CLI with `--env FOO=BAR` there are a few things that I don't understand after reading the CLI documentation:
1) Does `FOO` already need to exist in that project's environment variables (`vercel env add` or added through the dashboard)?
2) Is `FOO` set for all preview deployments or only for that specific deployment?
3) Is there a way to view the environment variables for a specific deployment? Because, I've tried the above and I don't see `FOO` in the project dashboard (we haven't created it as a resource for the project) and I don't see it in the build logs.

**Comment by amyegan** [comment, marked as answer]
Great questions @johnrichardrinehart! This type of per-deployment environment variable does work a little differently from the ones managed in your project dashboard.

> 1. Does FOO already need to exist in that project's environment variables (vercel env add or added through the dashboard)?

No, `FOO` does not need to exist in the project's environment variables. `FOO`, in this case, is a one-time variable that would need to be added for each deployment.

> 2. Is FOO set for all preview deployments or only for that specific deployment?

Variables added with [the `--env` or `--build-env` options](https://vercel.com/docs/cli/deploy#unique-options) are set only for that specific deployment. You would need to use [`vercel env add FOO preview`](https://vercel.com/docs/cli/env#extended-usage) to more permanently add a variable for all preview deployments.

> 3. Is there a way to view the environment variables for a specific deployment? Because, I've tried the above and I don't see FOO in the project dashboard (we haven't created it as a resource for the project) and I don't see it in the build logs.

Deployment-specific variables set with the `--env` or `--build-env` options are not added to the project dashboard. I ~~know~~ _don't know_ of a way to review variables for a specific deployment. If that's something you need, let me know and I'll make sure your feedback gets shared with the right team!

**Reply by johnrichardrinehart** [reply]
Thanks for your prompt response! Okay, a few follow-up questions.

1. So, `--env FOO=bar` should not overwrite `FOO` if it _does_ already exist as a project environment variable (for any environment), right? Just want to make extra sure.

2. Are any of the project-specific environment variables (if not provided explicitly with `--env X=Y` or `-e X=y` at deploy-time) sourced from the project environment variables as defaults? Or, are the only environment variables that are set/available those that are explicitly provided at deploy-time?

3. You said
> I know of a way to review variables for a specific deployment.

That's a typo, right? You _don't_ know of a way to review variables for a specific deployment.

**Reply by amyegan** [reply]
> 1. So, --env FOO=bar should not overwrite FOO if it does already exist as a project environment variable (for any environment), right? Just want to make extra sure.

If you already have a `FOO` variable set for your project, then `--env FOO=bar` will use the new value for this deployment only. It will not update the value in the project dashboard, so other deployments would still use that more permanent value set in the dashboard.

> 2. Are any of the project-specific environment variables (if not provided explicitly with --env X=Y or -e X=y at deploy-time) sourced from the project environment variables as defaults? Or, are the only environment variables that are set/available those that are explicitly provided at deploy-time?

Yes. If no `--env` or `--build-env` variables are set on the deployment, it will look to the project environment variables and use any applicable values found there. You are free to use a combination of the two if you need to override a single variable for a deployment.

> 3. That's a typo, right? You don't know of a way to review variables for a specific deployment.

Yes, that was absolutely a typo! Thanks for calling it out. I'll update the original comment for clarity 😆

**Reply by johnrichardrinehart** [reply]
Thanks, again! Okay, I promise we're almost done! One more follow-up question. Re
> Yes. If no --env or --build-env variables are set on the deployment, it will look to the project environment variables and use any applicable values found there. You are free to use a combination of the two if you need to override a single variable for a deployment.

Imagine that we have `FOO` set to different values for `production`, `preview`, `branch`-specific, and `development`. I'm guessing that if we don't provide `FOO` at deploy-time (no use of `--env`/`-e`, say) that either of the `branch`-specific and `development` values would/will not be used as the default value (because no branch deployment is special/source-of-truth and `development` makes no sense). But, we'd like to have the `production` environment variables pulled in, if a corresponding value is not defined on the command line as an argument (`--env`/`-e`). But, since these `vercel` CLI deployments look like `preview` builds in the dashboard (have `preview` label below the deployments), I'm guessing that the default values are the non-`branch`-specific `preview` environment variables. Is that right?

Also, regarding your offer to forward our request to view the build-time (`--build-env`) and runtime (`--env`) environment variables for a specific deployment (by deployment ID, say): Please! I'd love to have this visibility into how a specific deployment was built and should be running. I _guess_ we could log runtime environment variables or expose an endpoint that returns them in the response.... But, that doesn't sound good.

**Reply by amyegan** [reply]
If no deployment-specific variable value is provided, then the deployment will fall back to whichever variable matches in the project environment variables. Production deployments will only use production variables. Branch-specific preview variables will only be applied to preview deployments from the specified branch. If no matching branch-specific variable is found, it will use general preview variables. If no variable is found matching the environment, the value will be `undefined`.

**Reply by johnrichardrinehart** [reply]
Thanks!
